crates/bifrost/src/providers/replicated_loglet/tasks/check_seal.rs

```diff
@@ -0,0 +1,120 @@
+    ProbablyOpen,
+}
+
+impl CheckSealTask {
+    pub async fn run<T: TransportConnect>(
+        my_params: &ReplicatedLogletParams,
+        get_loglet_info_rpc: &RpcRouter<GetLogletInfo>,
+        known_global_tail: &TailOffsetWatch,
+        networking: &Networking<T>,
+    ) -> Result<CheckSealOutcome, ShutdownError> {
+        // If all nodes in the nodeset is in "provisioning", we can confidently short-circuit
+        // the result to LogletOffset::Oldest and the loglet is definitely unsealed.
+        if my_params
+            .nodeset
+            .all_provisioning(&networking.metadata().nodes_config_ref())
+        {
+            return Ok(CheckSealOutcome::ProbablyOpen);
+        }
+        // todo: If effective nodeset is empty, should we consider that the loglet is implicitly
+        // sealed?
+
+        let effective_nodeset = EffectiveNodeSet::new(
+            &my_params.nodeset,
+            &networking.metadata().nodes_config_ref(),
+        );
+
+        let mut nodeset_checker = NodeSetChecker::<'_, NodeTailStatus>::new(
+            &effective_nodeset,
+            &networking.metadata().nodes_config_ref(),
+            &my_params.replication,
+        );
+
+        let mut nodes = effective_nodeset.shuffle_for_reads(networking.metadata().my_node_id());
+
+        let mut cancel = std::pin::pin!(cancellation_watcher());
+        trace!(
+            loglet_id = %my_params.loglet_id,
+            effective_nodeset = %effective_nodeset,
+            "Checking seal status for loglet",
+        );
+        loop {
+            if nodeset_checker
+                .check_fmajority(NodeTailStatus::is_known_unsealed)
+                .passed()
+            {
+                // once we reach f-majority of unsealed, we stop.
+                return Ok(CheckSealOutcome::ProbablyOpen);
+            }
+
+            let Some(next_node) = nodes.pop() else {
+                info!(
+                    loglet_id = %my_params.loglet_id,
+                    effective_nodeset = %effective_nodeset,
+                    "Insufficient nodes responded to GetLogletInfo requests, we cannot determine seal status, we'll assume it's unsealed for now",
+                );
+                return Ok(CheckSealOutcome::ProbablyOpen);
+            };
+
+            let task = FindTailOnNode {
+                node_id: next_node,
+                loglet_id: my_params.loglet_id,
+                get_loglet_info_rpc,
+                known_global_tail,
+            };
+            let tail_status = tokio::select! {
+                _ = &mut cancel => {
+                    return Err(ShutdownError);
+                }
+                (_, tail_status) = task.run(networking) => { tail_status },
+            };
+            if tail_status.is_known_sealed() {
+                // we only need to see a single node sealed to declare that we are probably sealing (or
+                // sealed)
+                return Ok(CheckSealOutcome::ProbablyOpen);
```

Comment:
I guess this should return `CheckSealOutcome::Sealing`.

Reply:
Fix sent in #2094